lib/Optimizer/TrainingPreparation.cpp

```diff
@@ -0,0 +1,113 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Optimizer/TrainingPreparation.h"
+#include "glow/Graph/PlaceholderBindings.h"
+
+namespace glow {
+
+TensorInitializer getDefaultTensorInitializer() {
+  return [](Function *F, Node *node, NodeValue &input, Tensor *tensor) {
+    switch (node->getKind()) {
+    case Kinded::Kind::ConvolutionNodeKind: {
+      ConvolutionNode *CN = llvm::dyn_cast<ConvolutionNode>(node);
+      if (CN->getFilter() == input) {
+        ShapeNHWC idim = ShapeNHWC(CN->getInput().dims());
+        ShapeHW kdim(CN->getKernels());
+        size_t fanIn = kdim.height * kdim.width * idim.c;
+        tensor->init(Tensor::InitKind::Xavier, fanIn, F->getPRNG());
+      } else if (CN->getBias() == input) {
+        tensor->init(Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+      }
+    } break;
+    case Kinded::Kind::BatchNormalizationNodeKind: {
+      BatchNormalizationNode *BN = llvm::dyn_cast<BatchNormalizationNode>(node);
+      if (BN->getScale() == input) {
+        tensor->init(glow::Tensor::InitKind::Zero, 0, F->getPRNG());
+      } else if (BN->getBias() == input) {
+        tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+      } else if (BN->getMean() == input) {
+        tensor->zero();
+      } else if (BN->getVar() == input) {
+        tensor->init(glow::Tensor::InitKind::Broadcast, 1.0, F->getPRNG());
+      }
+    } break;
+    case Kinded::Kind::FullyConnectedNodeKind: {
+      FullyConnectedNode *FCN = llvm::dyn_cast<FullyConnectedNode>(node);
+      if (FCN->getWeights() == input) {
+        auto in = FCN->getInput();
+        tensor->init(glow::Tensor::InitKind::Zero, in.dims()[1], F->getPRNG());
+      } else if (FCN->getBias() == input) {
+        tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+      }
+    } break;
+    default:
+      break;
+    }
+  };
+}
+
+void prepareFunctionForTraining(
+    Function *F, PlaceholderBindings &bindings,
+    const std::unordered_set<std::string> &legalOutputNames,
+    TensorInitializer &&initializer) {
+
+  auto &nodes = F->getNodes();
+  // Lookup all nodes, skip Storage types, enumerate inputs,
+  // replace Constant type with trainable Placeholders except special cases,
+  // like BatchNormalization inputs (mean and variance). In special cases
+  // replace Constant type with non-trainable Placeholders.
+  for (auto &node : nodes) {
+    // Skip storages.
+    if (Storage::classof(&node)) {
```

Comment:
The more usual way how it is done in Glow and LLVM is `llvm::isa<Storage>(&node)`.

Reply:
Class method checks if node is Constant or Placeholder. I know at this moment there are only two derived classes but it could be changed in the future.

Reply:
I agree with @opti-mix. `llvm::isa<Storage>(&node)` will work even if more derived classes are added, right? What is the advantage of using `Storage::classof`?